crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -800,6 +800,123 @@ impl Client {
         None
     }
 
+    pub async fn state_diff_for_block(
+        self,
+        block: BlockNumber,
+        state_diff_length: u64,
+    ) -> Option<(PeerId, StateUpdateData)> {
+        let request = StateDiffsRequest {
+            iteration: Iteration {
+                start: block.get().into(),
+                direction: Direction::Forward,
+                limit: 1,
+                step: 1.into(),
+            },
+        };
+
+        let peers = self
+            .get_update_peers_with_sync_capability(protocol::StateDiffs::NAME)
+            .await;
+
+        'next_peer: for peer in peers {
+            let Ok(mut stream) = self
+                .inner
+                .send_state_diffs_sync_request(peer, request)
+                .await
+                .inspect_err(|error| tracing::debug!(%peer, %error, "State diffs request failed"))
+            else {
+                continue;
+            };
+
+            let mut current_count = state_diff_length;
+            let mut state_diff = StateUpdateData::default();
+
+            while let Some(resp) = stream.next().await {
+                match resp {
+                    StateDiffsResponse::ContractDiff(ContractDiff {
+                        address,
+                        nonce,
+                        class_hash,
+                        values,
+                        domain: _,
+                    }) => {
+                        match current_count.checked_sub(values.len().try_into().unwrap()) {
```

Comment:
I think you need to account for [nonce and class updates too](https://github.com/starknet-io/starknet-p2p-specs/blob/main/p2p/proto/protocols.md#length)